components/script/dom/location.rs

```diff
@@ -217,7 +241,20 @@ impl LocationMethods for Location {
     // https://html.spec.whatwg.org/multipage/#dom-location-protocol
     fn SetProtocol(&self, value: USVString) -> ErrorResult {
         self.check_same_origin_domain()?;
-        self.set_url_component(value, UrlHelper::SetProtocol);
+        let url = self.window.get_url().into_url();
+        // TODO: avoid double parse,
+        // restructure set_url_component to accomodate various checks.
+        let mut try_parse = ServoUrl::from_url(url);
+        UrlHelper::SetProtocol(&mut try_parse, value.clone());
```

Comment:
This looks like a different algorithm from `https://html.spec.whatwg.org/multipage/#dom-location-protocol`, does it achieve the same result?

Reply:
I think so, yes. The call to `self.set_url_component` would correspond to Step 7, and these changes attempt to introduce the checks made at 4 to 6, before going to step 7. 

The problem is the double call to `UrlHelper::SetProtocol`, and the need to copy the url to do the check since we don't want to mutate it as part of it. 

There is an issue addressing this more comprehensively at https://github.com/servo/servo/issues/23522

Reply:
OK, cite the issue and that's fine.